### ISSUE TITLE
[Bug]: min_tokens + structured output can empty the token mask and return HTTP 500

### ISSUE THREAD

**Issue by Yunzez** [vllm-project/vllm#53665]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
((vllm-19) ) gpu-a6-7:~$ python -m vllm.collect_env
Collecting environment information...
==============================
        System Info
==============================
OS                           : Red Hat Enterprise Linux release 8.10 (Ootpa) (x86_64)
GCC version                  : (GCC) 8.5.0 20210514 (Red Hat 8.5.0-28)
Clang version                : Could not collect
CMake version                : version 4.3.1
Libc version                 : glibc-2.28

==============================
       PyTorch Info
==============================
PyTorch version              : 2.10.0+cu128
Is debug build               : False
CUDA used to build PyTorch   : 12.8
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.12 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 20:16:04) [GCC 11.2.0] (64-bit runtime)
Python platform              : Linux-4.18.0-553.89.1.el8_10.x86_64-x86_64-with-glibc2.28

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   :
GPU models and configuration : GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version        : 590.44.01
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              96
On-line CPU(s) list: 0-95
Thread(s) per core:  1
Core(s) per socket:  48
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               143
Model name:          Intel(R) Xeon(R) Platinum 8468
Stepping:            8
CPU MHz:             3800.000
CPU max MHz:         3800.0000
CPU min MHz:         800.0000
BogoMIPS:            4200.00
Virtualization:      VT-x
L1d cache:           48K
L1i cache:           32K
L2 cache:            2048K
L3 cache:            107520K
NUMA node0 CPU(s):   0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94
NUMA node1 CPU(s):   1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.5.2
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-frontend==1.22.1
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-cufile-cu12==1.13.1.3
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cutlass-dsl==4.4.2
[pip3] nvidia-cutlass-dsl-libs-base==4.4.2
[pip3] nvidia-ml-py==13.595.45
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvshmem-cu12==3.4.5
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] pyzmq==27.1.0
[pip3] torch==2.10.0+cu128
[pip3] torchaudio==2.9.0
[pip3] torchvision==0.25.0+cu128
[pip3] transformers==4.57.6
[pip3] triton==3.6.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.19.0
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled
GPU Topology:
  	GPU0	NIC0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	NODE	0,2,4,6,8,10	0		N/A
NIC0	NODE	 X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0

==============================
     Environment Variables
==============================
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_user

((vllm-19) ) gpu-a6-7:~$
```

</details>

#### This appears to be the same underlying issue, seems to be addressed by #47489. I’m filing this because it is still reproducible on v0.27.1 through the stock OpenAI API with ordinary object schemas, where it deterministically returns HTTP 500. 

A request using both `min_tokens` and structured output can leave vLLM with **no legal token to sample** once the constrained value is complete but `min_tokens` has not yet been reached.
This appears to be the same underlying issue addressed by #47489. I’m filing this because it is still reproducible on v0.27.1 through the stock OpenAI API with ordinary object schemas, where it deterministically returns HTTP 500.
On vLLM 0.27.1, this deterministically causes the request to fail with HTTP 500. The same root cause was also present in 0.19.0, that's when I first found it, **back then it instead silently emitted token ID 0 until `min_tokens` was reached**.

This appears to be the same mechanism addressed by #47489. BUT, this report adds a normal multi-field JSON schema reproduction and confirms that JSON mode and regex constraints are affected as well.

## current environment recap

* vLLM: `0.27.1`
* Image: `vllm/vllm-openai:v0.27.1`
* Model: `Qwen/Qwen2.5-0.5B-Instruct`
* GPU: H100 80GB
* Structured-output backend: default `auto` → xgrammar
* Decoding: greedy (`temperature=0`)

No special structured-output configuration is required.

## Steps to reproduce

```bash
curl -s http://localhost:8000/v1/chat/completions \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "Qwen/Qwen2.5-0.5B-Instruct",
    "messages": [
      {"role": "user", "content": "Give me a name and a count."}
    ],
    "temperature": 0,
    "max_tokens": 96,
    "min_tokens": 64,
    "response_format": {
      "type": "json_schema",
      "json_schema": {
        "name": "flat",
        "schema": {
          "type": "object",
          "properties": {
            "name": {"type": "string"},
            "count": {"type": "integer"}
          },
          "required": ["name", "count"],
          "additionalProperties": false
        }
      }
    }
  }'
```

## Expected vs actual

**Expected:** The request returns schema-valid JSON. If the grammar completes before `min_tokens`, the grammar constraint should remain authoritative rather than producing an invalid sampling state.

**Actual on 0.27.1:** HTTP 500.

```text
Failed to advance FSM ... for tokens 0. Please file an issue.
Unexpected: grammar rejected tokens [0] ... Terminating request.
Request ... failed with an internal error during generation
```

The server remains healthy; only the request fails.

On v0.19.0, the same condition was silent instead: greedy sampling selected vocabulary index 0 repeatedly until `min_tokens` was reached, then EOS was sampled and the request returned HTTP 200 with invalid structured output.

## Root cause

The problem is the composition of two individually valid masks:

1. Once the structured value is complete, xgrammar allows only the stop token.
2. While the output is shorter than `min_tokens`, `MinTokensLogitsProcessor` masks the stop tokens with `-inf`.
3. The grammar mask is applied before the logits processor.
4. Their intersection is therefore empty: every logit becomes `-inf`.
5. Greedy sampling uses an unguarded `argmax`, which returns vocabulary index 0.

The newer grammar check catches token 0 after sampling, which turns the older silent corruption into the current HTTP 500. It does not prevent the empty mask.

Did a 24-configuration sweep, the failure occurs when `min_tokens` exceeds the number of tokens required to complete the constrained value. Adding a `stop` string does not change the result.

Guidance and lm-format-enforcer appear to have the same terminal-mask behavior, but I have not tested them. Outlines appears to behave differently and was also not tested.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by dhakshin32** [comment]
I can take this

**Comment by ActiveSky** [comment]
Hi @Yunzez and @dhakshin32, thank you for the detailed report and for offering to investigate it.

I’m sharing a quick check as a community contributor, not a maintainer: I could not find a GitHub-linked PR for #53665 in the issue timeline or by searching PR bodies. I did find earlier open work on the same underlying interaction between `min_tokens` and terminal structured-output grammars: [#47489](https://github.com/vllm-project/vllm/pull/47489) (currently showing merge conflicts) and draft [#49521](https://github.com/vllm-project/vllm/pull/49521). The issue description also points to #47489, so I think these are semantically related rather than PRs already attached to this issue.

Since @dhakshin32 has offered to take this, I don’t want to duplicate the effort. If useful, could the maintainers and contributors advise whether the new multi-field JSON-schema reproduction should be folded into one of those PRs, or whether a separate fix is needed?

Thank you.

**Comment by laulopezreal** [comment]
Looking into this. The root cause is clear: grammar completion allows only the stop token while MinTokensLogitsProcessor masks stop tokens until min_tokens is reached, so their intersection is empty and the unguarded argmax returns index 0. I'll open a fix with a regression test.

**Comment by LiRunGuo** [comment]
## Independent reproduction: 4/4 requests deterministically crash (vLLM v0.24.0 and main)

Reproduced the exact failure on **vLLM v0.24.0** (this run) and **vLLM main @ 94d96e2** (earlier run): `Qwen2.5-0.5B-Instruct`, `temperature=0.0`, `max_tokens=96`, `min_tokens=64`, `structured_outputs={json: <{"name": string, "count": integer} schema>}`. All 4 prompts in a single batch deterministically hit the crash:

```
ERROR [backend_xgrammar.py:162] Failed to advance FSM for request 1-9a368fa2 for tokens 0. Please file an issue.
ERROR [scheduler.py:1606] Unexpected: grammar rejected tokens [0] for request 1-9a368fa2. Terminating request.
```

(identical lines for requests 0, 2, 3). Every output finishes with `finish_reason=error` at 10–13 generated tokens — i.e. the grammar completes while `min_tokens=64` is still far away, exactly the scenario in the report.

### Minimal repro (offline mode, no server needed)

```python
from vllm import LLM, SamplingParams
from vllm.sampling_params import StructuredOutputsParams

llm = LLM(model="Qwen2.5-0.5B-Instruct", enforce_eager=True,
          max_model_len=2048, gpu_memory_utilization=0.35)

params = SamplingParams(
    temperature=0.0, max_tokens=96, min_tokens=64,
    structured_outputs=StructuredOutputsParams(json={
        "type": "object",
        "properties": {"name": {"type": "string"}, "count": {"type": "integer"}},
        "required": ["name", "count"],
        "additionalProperties": False,
    }),
)
outputs = llm.generate(["Give me a name and a count."] * 4, params)
# all outputs[].finish_reason == "error"
```

### Root-cause confirmation

Agrees with the analysis above (and with laulopezreal's): once the grammar is complete, the only legal token is the stop token, but `MinTokensLogitsProcessor` (`vllm/v1/sample/logits_processor/builtin.py:165`) keeps masking stop tokens until `min_tokens` is reached; the intersection becomes empty, argmax yields token 0, and the grammar rejects it (`scheduler.py:1606`). Both open PRs (#47489, #49521) target this; #47489 currently needs a rebase.

Happy to add any further detail or run additional configurations (e.g. other `min_tokens` values, `allowed_token_ids` interplay) if useful.

**Comment by vadiklyutiy** [comment]
should be fixed in #54218
